File: src/types/notifications.ts
/**
 * 通知系统 - 前后端通信数据定义
 * 用于头像旁的通知面板
 */

// ============= 通知优先级与类型 =============
export const NotificationType = {
  LEARNING_PROGRESS: 'learning_progress',      // 学习进度更新
  TASK_COMPLETED: 'task_completed',            // 任务完成
  PENDING_QUESTIONS: 'pending_questions',      // 待处理问题
  LEARNING_RECOMMENDATION: 'learning_recommendation', // 学习建议
  CONFUSION_DETECTED: 'confusion_detected',    // 卡顿检测
  ACHIEVEMENT_UNLOCKED: 'achievement_unlocked', // 成就解锁
  DEADLINE_REMINDER: 'deadline_reminder',      // 截止日期提醒
  RESOURCE_AVAILABLE: 'resource_available',    // 补充资源可用
  PEER_INTERACTION: 'peer_interaction',        // 同学互动
  SYSTEM_MESSAGE: 'system_message'             // 系统消息
} as const;

export type NotificationTypeKey = typeof NotificationType[keyof typeof NotificationType];

export const NotificationPriority = {
  URGENT: 'urgent',
  HIGH: 'high',
  NORMAL: 'normal',
  LOW: 'low'
} as const;

export type NotificationPriorityType = typeof NotificationPriority[keyof typeof NotificationPriority];

// ============= 单个通知对象 =============
export interface Notification {
  id: string;
  userId: string;
  
  // 基础信息
  type: NotificationTypeKey;
  title: string;
  message: string;
  priority: NotificationPriorityType;
  
  // 样式与呈现
  icon?: string;                    // 图标名称或 emoji
  backgroundColor?: string;         // 背景颜色（可选）
  
  // 操作相关
  actionUrl?: string;              // 点击后的跳转 URL
  actionLabel?: string;            // 按钮文本
  actionType?: 'navigate' | 'open_modal' | 'trigger_task' | 'none'; // 行为类型
  actionPayload?: Record<string, unknown>; // 行为参数
  
  // 详细数据
  metadata?: {
    documentId?: string;
    annotationId?: string;
    taskId?: string;
    projectId?: string;
    relatedContent?: string;       // 相关内容摘要
    
    // 进度相关
    currentProgress?: number;      // 0-100
    targetProgress?: number;
    
    // 问题相关
    questionCount?: number;
    completedCount?: number;
    
    // 时间相关
    estimatedTimeMinutes?: number;
    dueDate?: number;
    
    [key: string]: unknown;
  };
  
  // 时间信息
  createdAt: number;
  scheduledAt?: number;            // 计划发送时间
  expiresAt?: number;              // 过期时间
  
  // 状态
  isRead: boolean;
  readAt?: number;
  isDismissed?: boolean;
  dismissedAt?: number;
}

// ============= 通知数组包装 =============
export interface NotificationBatch {
  notifications: Notification[];
  totalCount: number;
  unreadCount: number;
  hasMore: boolean;
}

// ============= 前端 → 后端：获取通知 =============
export interface FetchNotificationsRequest {
  userId: string;
  projectId?: string;
  limit?: number;              // 默认 20
  offset?: number;             // 分页偏移，默认 0
  types?: NotificationTypeKey[]; // 过滤通知类型
  unreadOnly?: boolean;         // 只返回未读，默认 false
  sortBy?: 'recent' | 'priority'; // 排序方式，默认 'recent'
}

export interface FetchNotificationsResponse {
  status: 'success' | 'error';
  data?: NotificationBatch;
  error?: {
    code: string;
    message: string;
  };
}

// ============= 前端 → 后端：标记通知为已读 =============
export interface MarkNotificationReadRequest {
  notificationId: string;
  userId: string;
}

export interface MarkNotificationReadResponse {
  status: 'success' | 'error';
  notification?: Notification;
  error?: {
    code: string;
    message: string;
  };
}

// ============= 前端 → 后端：清除通知 =============
export interface DismissNotificationRequest {
  notificationId: string;
  userId: string;
}

export interface BatchDismissNotificationsRequest {
  notificationIds: string[];
  userId: string;
}

export interface DismissNotificationResponse {
  status: 'success' | 'error';
  dismissed: boolean;
  error?: {
    code: string;
    message: string;
  };
}

// ============= 后端 → 前端：实时推送（WebSocket 或轮询） =============
/**
 * 后端通过 WebSocket 或 Server-Sent Events 推送新通知
 * Tauri 中枢应该实现监听器
 */
export interface NotificationPushEvent {
  eventType: 'new_notification' | 'notification_updated' | 'notification_dismissed';
  notification: Notification;
  timestamp: number;
}

// ============= 具体场景的通知数据示例 =============

/**
 * 学习进度通知
 */
export interface LearningProgressNotification extends Notification {
  type: 'learning_progress';
  metadata?: {
    documentId: string;
    currentProgress: number;    // 0-100
    targetProgress: number;
    topicsLearned?: string[];
    nextTopic?: string;
  };
}

/**
 * 任务完成通知
 */
export interface TaskCompletedNotification extends Notification {
  type: 'task_completed';
  metadata?: {
    taskId: string;
    taskName: string;
    completionTime: number;     // 完成耗时（秒）
    points?: number;            // 获得的积分
    nextTask?: {
      id: string;
      name: string;
    };
  };
}

/**
 * 卡顿检测通知
 */
export interface ConfusionDetectedNotification extends Notification {
  type: 'confusion_detected';
  metadata?: {
    documentId: string;
    annotationId: string;
    topic: string;
    detectedAt: number;
    timeStuckSeconds: number;
    suggestedActions?: Array<{
      action: string;
      description: string;
    }>;
    recommendedResources?: Array<{
      id: string;
      title: string;
      type: 'video' | 'article' | 'exercise';
      url: string;
    }>;
  };
}

/**
 * 待处理问题通知
 */
export interface PendingQuestionsNotification extends Notification {
  type: 'pending_questions';
  metadata?: {
    questionCount: number;
    completedCount: number;
    topicsWithQuestions?: string[];
    deadlineTimestamp?: number;
  };
}

/**
 * 成就解锁通知
 */
export interface AchievementUnlockedNotification extends Notification {
  type: 'achievement_unlocked';
  metadata?: {
    achievementId: string;
    achievementName: string;
    achievementIcon?: string;
    points: number;
    description: string;
    nextAchievement?: {
      name: string;
      progress: number; // 0-100
    };
  };
}

// ============= 后端配置：通知规则 =============
/**
 * 用户的通知偏好设置（可存储在 localStorage 或后端）
 */
export interface NotificationPreferences {
  userId: string;
  
  // 启用的通知类型
  enabledTypes: NotificationTypeKey[];
  
  // 通知方式
  enableBrowserNotifications: boolean;
  enableInAppNotifications: boolean;
  enableEmailNotifications: boolean;
  
  // 优先级过滤
  minPriorityLevel: NotificationPriorityType;
  
  // 时间设置
  quietHours?: {
    enabled: boolean;
    from: string;  // "22:00"
    to: string;    // "08:00"
  };
  
  // 频率控制
  maxNotificationsPerHour?: number;
  
  updatedAt: number;
}

// ============= 后端 API 路由定义 =============
/**
 * 后端应该提供的通知相关 API 端点：
 * 
 * GET    /api/notifications                    - 获取通知列表
 * POST   /api/notifications/:id/read           - 标记为已读
 * DELETE /api/notifications/:id                - 删除通知
 * POST   /api/notifications/batch-dismiss      - 批量删除
 * GET    /api/notifications/preferences        - 获取偏好设置
 * PUT    /api/notifications/preferences        - 更新偏好设置
 * POST   /api/notifications/test               - 发送测试通知（仅开发）
 * 
 * WebSocket 路由：
 * WS     /ws/notifications/:userId             - 实时通知推送
 */

File: src-tauri/src/notifications.rs
/// 通知系统 - Tauri 命令处理器
/// 
/// 实现前端与 Python 后端的通知系统通信

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use reqwest::Client;
use tauri::State;

// ============= 数据结构 =============

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Notification {
    pub id: String,
    #[serde(rename = "userId")]
    pub user_id: String,
    #[serde(rename = "type")]
    pub notification_type: String,
    pub title: String,
    pub message: String,
    pub priority: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "actionUrl")]
    pub action_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "actionLabel")]
    pub action_label: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "actionType")]
    pub action_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "actionPayload")]
    pub action_payload: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<HashMap<String, serde_json::Value>>,
    #[serde(rename = "createdAt")]
    pub created_at: i64,
    #[serde(skip_serializing_if = "Option::is_none", rename = "scheduledAt")]
    pub scheduled_at: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "expiresAt")]
    pub expires_at: Option<i64>,
    #[serde(rename = "isRead")]
    pub is_read: bool,
    #[serde(skip_serializing_if = "Option::is_none", rename = "readAt")]
    pub read_at: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "isDismissed")]
    pub is_dismissed: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "dismissedAt")]
    pub dismissed_at: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NotificationBatch {
    pub notifications: Vec<Notification>,
    #[serde(rename = "totalCount")]
    pub total_count: i32,
    #[serde(rename = "unreadCount")]
    pub unread_count: i32,
    #[serde(rename = "hasMore")]
    pub has_more: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct NotificationPreferences {
    #[serde(rename = "userId")]
    pub user_id: String,
    #[serde(rename = "enabledTypes")]
    pub enabled_types: Vec<String>,
    #[serde(rename = "enableBrowserNotifications")]
    pub enable_browser_notifications: bool,
    #[serde(rename = "enableInAppNotifications")]
    pub enable_in_app_notifications: bool,
    #[serde(rename = "enableEmailNotifications")]
    pub enable_email_notifications: bool,
    #[serde(rename = "minPriorityLevel")]
    pub min_priority_level: String,
    #[serde(skip_serializing_if = "Option::is_none", rename = "quietHours")]
    pub quiet_hours: Option<QuietHours>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "maxNotificationsPerHour")]
    pub max_notifications_per_hour: Option<i32>,
    #[serde(rename = "updatedAt")]
    pub updated_at: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct QuietHours {
    pub enabled: bool,
    pub from: String,
    pub to: String,
}

pub struct AppState {
    pub http_client: Client,
    pub backend_url: String,
}

// ============= Tauri 命令 =============

#[tauri::command]
pub async fn fetch_notifications(
    state: State<'_, AppState>,
    user_id: String,
    project_id: Option<String>,
    limit: u32,
    offset: u32,
    types: Vec<String>,
    unread_only: bool,
    sort_by: String,
) -> Result<NotificationBatch, String> {
    let url = format!("{}/api/notifications/fetch", state.backend_url);
    
    let body = serde_json::json!({
        "user_id": user_id,
        "project_id": project_id,
        "limit": limit,
        "offset": offset,
        "types": types,
        "unread_only": unread_only,
        "sort_by": sort_by,
    });
    
    let response = state.http_client
        .post(&url)
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    response.json::<NotificationBatch>()
        .await
        .map_err(|e| format!("INVALID_RESPONSE: {}", e))
}

#[tauri::command]
pub async fn mark_notification_as_read(
    state: State<'_, AppState>,
    notification_id: String,
    user_id: String,
) -> Result<Notification, String> {
    let url = format!("{}/api/notifications/{}/read", state.backend_url, notification_id);
    
    let body = serde_json::json!({
        "user_id": user_id,
    });
    
    let response = state.http_client
        .post(&url)
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if response.status() == 404 {
        return Err("NOT_FOUND: Notification not found".to_string());
    }
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    response.json::<Notification>()
        .await
        .map_err(|e| format!("INVALID_RESPONSE: {}", e))
}

#[tauri::command]
pub async fn dismiss_notification(
    state: State<'_, AppState>,
    notification_id: String,
    user_id: String,
) -> Result<serde_json::Value, String> {
    let url = format!("{}/api/notifications/{}", state.backend_url, notification_id);
    
    let response = state.http_client
        .delete(&url)
        .query(&[("user_id", user_id)])
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    response.json()
        .await
        .map_err(|e| format!("INVALID_RESPONSE: {}", e))
}

#[tauri::command]
pub async fn batch_dismiss_notifications(
    state: State<'_, AppState>,
    notification_ids: Vec<String>,
    user_id: String,
) -> Result<serde_json::Value, String> {
    let url = format!("{}/api/notifications/batch-dismiss", state.backend_url);
    
    let body = serde_json::json!({
        "notification_ids": notification_ids,
        "user_id": user_id,
    });
    
    let response = state.http_client
        .post(&url)
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    response.json()
        .await
        .map_err(|e| format!("INVALID_RESPONSE: {}", e))
}

#[tauri::command]
pub async fn get_notification_preferences(
    state: State<'_, AppState>,
    user_id: String,
) -> Result<NotificationPreferences, String> {
    let url = format!("{}/api/notifications/preferences", state.backend_url);
    
    let response = state.http_client
        .get(&url)
        .query(&[("user_id", user_id)])
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    response.json::<NotificationPreferences>()
        .await
        .map_err(|e| format!("INVALID_RESPONSE: {}", e))
}

#[tauri::command]
pub async fn update_notification_preferences(
    state: State<'_, AppState>,
    preferences: NotificationPreferences,
) -> Result<serde_json::Value, String> {
    let url = format!("{}/api/notifications/preferences", state.backend_url);
    
    let response = state.http_client
        .put(&url)
        .json(&preferences)
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    response.json()
        .await
        .map_err(|e| format!("INVALID_RESPONSE: {}", e))
}

#[tauri::command]
pub async fn refresh_notification_stream(
    state: State<'_, AppState>,
    user_id: String,
) -> Result<serde_json::Value, String> {
    // 刷新通知流 - 简单实现为重新获取通知
    let url = format!("{}/api/notifications/fetch", state.backend_url);
    
    let body = serde_json::json!({
        "user_id": user_id,
        "limit": 20,
        "offset": 0,
        "types": [],
        "unread_only": false,
        "sort_by": "recent",
    });
    
    let response = state.http_client
        .post(&url)
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    Ok(serde_json::json!({"success": true}))
}

#[tauri::command]
pub async fn send_test_notification(
    state: State<'_, AppState>,
    user_id: String,
    notification_type: String,
) -> Result<Notification, String> {
    let url = format!("{}/api/notifications/test", state.backend_url);
    
    let body = serde_json::json!({
        "user_id": user_id,
        "notification_type": notification_type,
    });
    
    let response = state.http_client
        .post(&url)
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("NETWORK_ERROR: {}", e))?;
    
    if response.status() == 403 {
        return Err("FORBIDDEN: Test notifications disabled".to_string());
    }
    
    if !response.status().is_success() {
        return Err(format!("SERVER_ERROR: {}", response.status()));
    }
    
    response.json::<Notification>()
        .await
        .map_err(|e| format!("INVALID_RESPONSE: {}", e))
}
